pytorch_lightning/trainer/supporters.py

```diff
@@ -328,12 +365,45 @@ def _calc_num_batches(loaders) -> int:
 
         raise TypeError(f'Got Type {type(all_lengths).__name__}, but expected one of Sequence, int or Mapping')
 
+    def __len__(self):
+        return self._calc_num_batches(self.loaders)
+
+
+class CombinedLoaderIterator(object):
+    """
+    Custom Iterator returning data from multple loaders, and allows sampling in parallel
+    """
+    def __init__(self, loaders: Any):
+        """
+
+        Args:
+            loaders: the loaders to sample from. Can be all kind of collection
+        """
+        self.loaders = loaders
+        self._loader_iters = None
+
+    @property
+    def loader_iters(self) -> Any:
+        if self._loader_iters is None:
+            self._loader_iters = self.create_loader_iters(self.loaders)
+
+        return self._loader_iters
+
+    def __iter__(self) -> Any:
+        return self
+
+    def __next__(self) -> Any:
+        if self._loader_iters is None:
+            self._loader_iters = self.create_loader_iters(self.loaders)
+        return self.request_next_batch(self._loader_iters)
```

Comment:
this does not support any cycling, but only uses the minimal length, since it does not handle` StopIteration` properly

Reply:
The `self.loaders` become cycling in `_wrap_loaders_max_size_cycle()` in `CombinedLoader`.